badge-maker/lib/badge-renderers.js

```diff
@@ -576,78 +580,111 @@ function forTheBadge({
-    // character.
-    labelWidth += 10 + 2 * label.length
-  } else if (hasLogo) {
+  const outLabelColor = labelColor || '#555'
+
+  // Compute text width.
+  // TODO: This really should count the symbols rather than just using `.length`.
+  // https://mathiasbynens.be/notes/javascript-unicode
+  // This is not using `preferredWidthOf()` as it tends to produce larger
+  // inconsistencies in the letter spacing. The badges look fine, however if you
+  // replace `textLength` with `letterSpacing` in the rendered SVG, you can see
+  // the discrepancy. Ideally, swapping out `textLength` for `letterSpacing`
+  // should not affect the appearance.
+  const labelTextWidth = label.length
+    ? (anafanafo(label, { font: `${FONT_SIZE}px Verdana` }) | 0) +
+      LETTER_SPACING * label.length
+    : 0
+  const messageTextWidth = message.length
+    ? (anafanafo(message, { font: `bold ${FONT_SIZE}px Verdana` }) | 0) +
+      LETTER_SPACING * message.length
+    : 0
+
+  // Compute horizontal layout.
+  // If a `labelColor` is set, the logo is always set against it, even when
+  // there is no label. When `needsLabelRect` is true, render a label rect and a
+  // message rect; when false, only a message rect.
+  const hasLabel = Boolean(label.length)
+  const needsLabelRect = hasLabel || (logo && labelColor)
+  let logoMinX, labelTextMinX
+  if (logo) {
+    logoMinX = LOGO_MARGIN
+    labelTextMinX = logoMinX + logoWidth + LOGO_TEXT_GUTTER
+  } else {
+    labelTextMinX = TEXT_MARGIN
+  }
+  let labelRectWidth, messageTextMinX, messageRectWidth
+  if (needsLabelRect) {
     if (hasLabel) {
-      labelWidth += 7
+      labelRectWidth = labelTextMinX + labelTextWidth + TEXT_MARGIN
     } else {
-      labelWidth -= 7
+      labelRectWidth = 2 * LOGO_MARGIN + logoWidth
     }
+    messageTextMinX = labelRectWidth + TEXT_MARGIN
+    messageRectWidth = 2 * TEXT_MARGIN + messageTextWidth
   } else {
-    labelWidth -= 11
+    if (logo) {
+      messageTextMinX = TEXT_MARGIN + logoWidth + LOGO_TEXT_GUTTER
+      messageRectWidth =
+        2 * TEXT_MARGIN + logoWidth + LOGO_TEXT_GUTTER + messageTextWidth
+    } else {
+      messageTextMinX = TEXT_MARGIN
+      messageRectWidth = 2 * TEXT_MARGIN + messageTextWidth
+    }
   }
 
-  // Add 20 px of padding, plus approximately 1.5 px of letter spacing per
-  // character.
-  messageWidth += 20 + 1.5 * message.length
-  const leftWidth = hasLogo && !hasLabel ? 0 : labelWidth
-  const rightWidth =
-    hasLogo && !hasLabel ? messageWidth + labelWidth : messageWidth
-
-  labelColor = hasLabel || hasLogo ? labelColor : color
-
-  color = escapeXml(color)
-  labelColor = escapeXml(labelColor)
-
-  let [leftLink, rightLink] = links
-  leftLink = escapeXml(leftLink)
-  rightLink = escapeXml(rightLink)
-  const { hasLeftLink, hasRightLink } = hasLinks({ links })
-
-  const accessibleText = createAccessibleText({ label, message })
+  // Render.
+  function renderLogo() {
```

Comment:
Do we actually need to define a second `renderLogo()` function here? It looks like we could just call the renderLogo() function we already have with the params

```js
renderLogo({
  logo,
  badgeHeight: BADGE_HEIGHT,
  horizPadding: logoMinX,
  logoWidth,
  logoPadding,
})
```

and it would return the same string?

Reply:
Just took another look at this.

The existing `renderLogo()` function we've got has three return values, and a bunch of parameters which I can't even quite make sense of. The new one I've written here is easier to understand. It's much clearer in for-the-badge, and I suspect will also make the other badge styles clearer.

Swapping in this implementation for the others would require adjusting the rest of the badges, which depend on the spaghetti logic of the old `renderLogo()`. I'd prefer to keep this refactor isolated to for-the-badge, with the idea that it can be used as a model to refactor the others.

How about I rename the existing `renderLogo()` to `renderLogoOld()`, and move the new function alongside it, so that it's ready to use in the other styles?

Reply:
Yeah your existing `renderLogo()` does more than one thing. The actual "render a string `<image...`" bit and the params read fairly interchangably in the two functions to me :shrug: . If you want to extract this `renderLogo()` as suggested and call the other one `renderLogoOld()`, you might as well have the "old" one call the new one if possible so you've got something like


```js
function renderLogo(params) {
  return `<image...`
}


function renderLogoOld({
  logo,
  badgeHeight,
  horizPadding,
  logoWidth = 14,
  logoPadding = 0,
}) {
  if (logo) {
    return {
      hasLogo: true,
      totalLogoWidth: logoWidth + logoPadding,
      renderedLogo: renderLogo(params),
    }
  } else {
    return { hasLogo: false, totalLogoWidth: 0, renderedLogo: '' }
  }
}
```

Once you've pulled it out of `forTheBadge()` I think they're going to end up with pretty much the same arg list.